### PR TITLE
use full_chunk queue in local shuffle and multiple sources in local passthrough

### PR DESCRIPTION
**Pull request by ZiheLiu**
If `_row_count >= _max_row_count`, all the sink operators sharing one `LocalExchangeMemoryManager` will be full.

- To make sure at least one partition source operator is ready to output chunk before all the sink operators are full, `max_row_count` must be at least `shuffle_partitions_num * config::vector_chunk_size`.
- The  local shuffle partitions may be slope, so a local shuffle source operator maybe receive multiple `full_chunk`. Therefore, use `queue _full_chunks` instead of one `_full_chunk`.
- Make local passthrough sink operator can output chunks to multiple exchange source operator. In this way, DOP of the source operator in `gather_pipelines_to_one` can be greater than one.

### CODE REVIEW

**Comment by kangkaisen** [comment]
> The local shuffle partitions may be slope, so a local shuffle source operator maybe receive multiple full_chunk. Therefore, use queue _full_chunks instead of one _full_chunk.

Use queue _full_chunks couldn't solve data skew， and will use more memory, What's the advantage of use queue _full_chunks ?

**Comment by kangkaisen** [comment]
> The local shuffle partitions may be slope, so a local shuffle source operator maybe receive multiple full_chunk. Therefore, use queue _full_chunks instead of one _full_chunk.

Use queue _full_chunks couldn't solve data skew， and will use more memory, What's the advantage of use queue _full_chunks ?

File: be/src/exec/pipeline/pipeline_builder.cpp [hunk 1]
```diff
@@ -64,11 +64,14 @@ OpFactories PipelineBuilderContext::maybe_interpolate_local_shuffle_exchange(
 }
 
 OpFactories PipelineBuilderContext::gather_pipelines_to_one(std::vector<OpFactories>& pred_operators_list) {
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_builder.cpp:66]: What' the difference `gather_pipelines_to_one` with `maybe_interpolate_local_passthrough_exchange `

**Comment by ZiheLiu** [comment]
> > The local shuffle partitions may be slope, so a local shuffle source operator maybe receive multiple full_chunk. Therefore, use queue _full_chunks instead of one _full_chunk.
> 
> Use queue _full_chunks couldn't solve data skew， and will use more memory, What's the advantage of use queue _full_chunks ?

@kangkaisen Because `max_row_count` must be at least `shuffle_partitions_num * config::vector_chunk_size` and data may be skew, a partition source maybe receives multiple full chunks. In this situation, a single `full_chunk` cannot receive multiple full chunks.

File: be/src/exec/pipeline/pipeline_builder.cpp (hunk 1 above)
Comment (ZiheLiu) [reply be/src/exec/pipeline/pipeline_builder.cpp:66]: `maybe_interpolate_local_passthrough_exchange()` merge outputs of multiple **drivers** of one pred pipeline into a driver of the post pipeline. It is used when the post operator cannot be parallelized and the pred operator can be parallelized. If the post operator and pred operator in a same pipeline, then the DOP of this pipeline can only be one. By using this method, we can divide the pred and post operator into two different pipelines.

`gather_pipelines_to_one()` merge outputs of multiple pred **pipeline** into a post pipeline. For example, UnionNode receives data from multiple child nodes, which belong to different pipelines.

**Comment by ZiheLiu** [comment]
run starrocks_clang-format

File: be/src/exec/pipeline/exchange/local_exchange.h [hunk 2]
```diff
@@ -101,12 +101,15 @@ class PassthroughExchanger final : public LocalExchanger {
...
-            _source->get_sources()[0]->finish(state);
+            for (auto* source : _source->get_sources()) {
+                source->finish(state);
+            }
         }
     }
 
 private:
     LocalExchangeSourceOperatorFactory* _source;
+    size_t _next_accept_source = 0;
```
Comment (satanson) [inline be/src/exec/pipeline/exchange/local_exchange.h:112]: use atomic variable，_next_accept_source is accessed by multiple threads.

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp [hunk 3]
```diff
@@ -9,16 +9,9 @@ namespace starrocks::pipeline {
 
 Status LocalExchangeSourceOperator::add_chunk(vectorized::ChunkPtr chunk) {
```
Comment (satanson) [inline be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp:10]: explain why  no merge?

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp [hunk 4]
```diff
@@ -31,28 +24,44 @@ Status LocalExchangeSourceOperator::add_chunk(vectorized::Chunk* chunk, const ui
     }
 
     if (_partial_chunk->num_rows() + size > config::vector_chunk_size) {
-        _full_chunk = std::move(_partial_chunk);
+        _full_chunks.emplace(std::move(_partial_chunk));
         _partial_chunk = chunk->clone_empty_with_slot();
     }
 
     _partial_chunk->append_selective(*chunk, indexes, from, size);
```
Comment (satanson) [inline be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp:31]: time-consuming operations, put outside of CS.

**Review by satanson — DISMISSED** [review]

**Review by Pslydhh — APPROVED** [review]

**Review by Pslydhh — DISMISSED** [review]

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp [hunk 5]
```diff
@@ -7,52 +7,90 @@
...
+               rows_num + _partition_chunk_queue.front().size <= config::vector_chunk_size) {
+            rows_num += _partition_chunk_queue.front().size;
+            selected_partition_chunks.emplace_back(std::move(_partition_chunk_queue.front()));
+            _partition_chunk_queue.pop();
+        }
+        _partition_rows_num -= rows_num;
+    }
+
+    // Unlock during merging partition chunks into a full chunk.
+    vectorized::ChunkPtr chunk = selected_partition_chunks[0].chunk->clone_empty_with_slot();
```
Comment (satanson) [inline be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp:86]: just append selected_partition_chunks[1..] to selected_partition_chunks[0]

File: be/src/exec/pipeline/exchange/local_exchange.cpp [hunk 6]
```diff
@@ -7,95 +7,107 @@
...
-            _channel_row_idx_start_points[i] += _channel_row_idx_start_points[i - 1];
-        }
+    auto& partitioner = _partitioners[sink_driver_sequence];
 
-        for (int i = num_rows - 1; i >= 0; --i) {
-            _row_indexes[_channel_row_idx_start_points[_hash_values[i]] - 1] = i;
-            _channel_row_idx_start_points[_hash_values[i]]--;
-        }
-    }
+    std::shared_ptr<std::vector<uint32_t>> partition_row_indexes = std::make_shared<std::vector<uint32_t>>(num_rows);
```
Comment (satanson) [inline be/src/exec/pipeline/exchange/local_exchange.cpp:73]: put partition_row_indexes into partitoiner, too, re-use it

**Review by kangkaisen — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Pull request merged**